src/bin/vhost_user_net.rs

```diff
@@ -244,52 +168,9 @@ impl VhostUserNetBackend {
             }
             return Ok(false);
         }
+        let write_complete = self.rx.process_desc_chain(&mem, next_desc, &mut queue);
 
-        // We just checked that the head descriptor exists.
-        let head_index = next_desc.as_ref().unwrap().index;
-        let mut write_count = 0;
-
-        // Copy from frame into buffer, which may span multiple descriptors.
-        loop {
-            match next_desc {
-                Some(desc) => {
-                    if !desc.is_write_only() {
-                        break;
-                    }
-                    let limit = cmp::min(write_count + desc.len as usize, self.rx.bytes_read);
-                    let source_slice = &self.rx.frame_buf[write_count..limit];
-                    let write_result = mem.write_slice(source_slice, desc.addr);
-
-                    match write_result {
-                        Ok(_) => {
-                            write_count = limit;
-                        }
-                        Err(e) => {
-                            error!("Failed to write slice: {:?}", e);
-                            break;
-                        }
-                    };
-
-                    if write_count >= self.rx.bytes_read {
-                        break;
-                    }
-                    next_desc = desc.next_descriptor();
-                }
-                None => {
-                    warn!("Receiving buffer is too small to hold frame of current size");
-                    break;
-                }
-            }
-        }
-
-        vring
-            .mut_queue()
-            .add_used(&mem, head_index, write_count as u32);
-
-        // Mark that we have at least one pending packet and we need to interrupt the guest.
-        self.rx.deferred_irqs = true;
-
-        Ok(write_count >= self.rx.bytes_read)
```

Comment:
Nice, thanks for this clean refactoring, we can clearly identify the benefit of factorization here.